src/Playwright/Core/WebSocketRoute.cs

```diff
@@ -0,0 +1,236 @@
+/*
+ * MIT License
+ *
+ * Copyright (c) Microsoft Corporation.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Text.Json;
+using System.Threading.Tasks;
+using Microsoft.Playwright.Helpers;
+using Microsoft.Playwright.Transport;
+using Microsoft.Playwright.Transport.Protocol;
+
+namespace Microsoft.Playwright.Core;
+
+/// <summary>
+/// <see cref="IWebSocketRoute"/>.
+/// </summary>
+internal class WebSocketRoute : ChannelOwner, IWebSocketRoute
+{
+    private readonly IWebSocketRoute _server;
+    internal readonly WebSocketRouteInitializer _initializer;
+    private bool _connected;
+    private Action<int?, string> _onPageClose;
+    private Action<IWebSocketFrame> _onPageMessage;
+    internal Action<int?, string> _onServerClose;
+    internal Action<IWebSocketFrame> _onServerMessage;
+
+    internal WebSocketRoute(ChannelOwner parent, string guid, WebSocketRouteInitializer initializer) : base(parent, guid)
+    {
+        _initializer = initializer;
+        MarkAsInternalType();
+        _server = new ServerWebSocketRoute(this);
+    }
+
+    public string Url => _initializer.Url;
+
+    internal override void OnMessage(string method, JsonElement? serverParams)
+    {
+        switch (method)
+        {
+            case "messageFromPage":
+                if (_onPageMessage != null)
+                {
+                    var frame = new WebSocketFrame(serverParams.Value.GetProperty("message").GetString(), serverParams.Value.GetProperty("isBase64").GetBoolean());
+                    _onPageMessage(frame);
```

Comment:
It seems like all the snippets for java/csharp should be updated to use `WebSocketFrame` instead of a string?

Reply:
Filed as https://github.com/microsoft/playwright/issues/33199.